[FLINK-38742][cdc-pipeline/postgres] Fix timestamp with time zone

Fluder-Paradyne opened the pull request:
In postgres when a table has a column with timestamp with time zone column the cdc pipeline fails

Mrart commented:
@Fluder-Paradyne  Thank for your contribute.
But I have some different opinions on the https://debezium.io/documentation/reference/stable/connectors/postgresql.html#postgresql-data-types in the document, “A string representation of a timestamp with timezone information,  where the timezone is GMT. "TIMESTAMPTZ means the time in which the GMT zone is stored, and the difference between timestampTZ and timestampLTZ is that TimestampLTZ stores the time in the GMT zone. I understand this place should be TimestampLTZ more reasonable?

cc @lvyanquan  Do you agree with me?

loserwang1024 reviewed (commented):
I have left some minor comment

Fluder-Paradyne committed:
[FLINK-38742][cdc-pipeline/postgres] Fix TIMESTAMPTZ type mapping to TIMESTAMP_LTZ

PostgreSQL TIMESTAMPTZ columns were causing NumberFormatException because
PostgresTypeUtils incorrectly mapped them to ZonedTimestampType (TIMESTAMP_WITH_TIME_ZONE).

PostgreSQL TIMESTAMPTZ stores values internally in UTC and converts on display
based on session timezone, which semantically matches TIMESTAMP_LTZ, not
TIMESTAMP_WITH_TIME_ZONE. Additionally, Debezium's PostgreSQL connector always
converts TIMESTAMPTZ values to UTC format (e.g., '2025-12-30T05:59:50.724893Z')
before serialization.

The type mismatch caused sinks to call getZonedTimestamp() on LocalZonedTimestampData,
resulting in NumberFormatException when trying to parse binary data as a string.

This commit fixes PostgresTypeUtils to correctly map TIMESTAMPTZ and TIMESTAMPTZ_ARRAY
to TIMESTAMP_LTZ type, matching both PostgreSQL semantics and Debezium's serialization
format. The E2E test is updated to verify the fix.

Fluder-Paradyne commented:
> @Fluder-Paradyne Thank for your contribute. But I have some different opinions on the https://debezium.io/documentation/reference/stable/connectors/postgresql.html#postgresql-data-types in the document, “A string representation of a timestamp with timezone information, where the timezone is GMT. "TIMESTAMPTZ means the time in which the GMT zone is stored, and the difference between timestampTZ and timestampLTZ is that TimestampLTZ stores the time in the GMT zone. I understand this place should be TimestampLTZ more reasonable?
> 
> cc @lvyanquan Do you agree with me?

Hey @Mrart thanks for the review, dug around a bit,  yes TIMESTAMP_LTZ is more reasonable as it close to postgres secmantics and sinks also support it

loserwang1024 commented:
@Fluder-Paradyne , I have seen your reproduce step.

```java
Caused by: java.lang.NumberFormatException: For input string: 
	at java.base/java.lang.NumberFormatException.forInputString(Unknown Source)
	at java.base/java.lang.Long.parseLong(Unknown Source)
	at java.base/java.lang.Long.parseLong(Unknown Source)
	at org.apache.flink.cdc.common.data.binary.BinaryRecordData.getZonedTimestamp(BinaryRecordData.java:182)
	at org.apache.flink.cdc.common.data.RecordData.lambda$createFieldGetter$f6fd429a$3(RecordData.java:233)
```

It seems that the type mapping is no problem. The root reasion is BinaryRecordData.getZonedTimestamp has problem.

Fluder-Paradyne commented:
@loserwang1024 for pg TIMESTAMPTZ column, the moment you insert/update the timezone information is converted to UTC and store as UTC time, and then when you select it, it gets converted based on timezone of the postgres session. 

Since the source itself does not have the original timezone info, shoudn't Timestamp ltz be fine ?

lvyanquan commented:
> Since the source itself does not have the original timezone info, shoudn't Timestamp ltz be fine ?

I agree with @loserwang1024's option, we can configure this information of Source through server-time-zone, refer to https://github.com/apache/flink-cdc/blob/533072b6bbbc266df7824b47433682885c123948/flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/options/JdbcSourceOptions.java#L67.

cloneot commented:
Is there any update?

fjanicki commented:
Any update? We encountered the same issue, making Flink unuseable with Postgres with timestamp columns. This is a pretty major bug. Can confirm that this PR fixed it for us by running a forked version with this patch applied. Anything blocking the upstream fix?

loserwang1024 commented:
> I am not sure if server-time-zone affects TIMESTAMPTZ.

@Fluder-Paradyne 
No influence. I have see the code of debezium. In Debezium's PostgreSQL connector, timestamp with time zone values are always normalized to UTC+0 regardless of the data source path. Here's why:

### Streaming Path (pgoutput logical replication)
When consuming WAL via the pgoutput plugin, the value goes through `AbstractColumnValue.asOffsetDateTimeAtUtc()` :
```java
return DateTimeFormat.get()
    .timestampWithTimeZoneToOffsetDateTime(asString())   // Step 1: parse raw text with original offset
    .withOffsetSameInstant(ZoneOffset.UTC);              // Step 2: normalize to UTC+0
```


The pgoutput protocol delivers column values as UTF-8 text strings (e.g., "2019-10-02T13:51:30.123456+02:00"). The parser first produces an OffsetDateTime with the original timezone offset from the WAL entry, then explicitly converts it to UTC via .withOffsetSameInstant(ZoneOffset.UTC). This guarantees the OffsetDateTime entering convertTimestampWithZone() is always at offset zero.

### Snapshot Path (JDBC)
During initial snapshots, the JDBC driver returns java.sql.Timestamp (a subclass of java.util.Date). In convertTimestampWithZone():
```java
if (data instanceof java.util.Date) {
    data = OffsetDateTime.ofInstant(((Date) data).toInstant(), ZoneOffset.UTC);
}
```
<img width="796" height="538" alt="image" src="https://github.com/user-attachments/assets/0af0542e-52d1-46cc-942d-3225c2906b64" />

java.util.Date.toInstant() yields an absolute point in time (epoch-based), and OffsetDateTime.ofInstant(..., ZoneOffset.UTC) renders it explicitly at UTC+0.


Thus, using timestamp ltz is ok because time offset is  normalized to UTC+0 by debezium. The only concern is compatibility, but it should be fine. Although the downstream jobs created previously used timestamp with time zone instead of timestamp with local time zone, no data has been written yet. Users will need to delete the downstream jobs and re-submit them. @leonardBang  ,WDYT?

leonardBang commented:
> > I am not sure if server-time-zone affects TIMESTAMPTZ.
> 
> @Fluder-Paradyne No influence. I have see the code of debezium. In Debezium's PostgreSQL connector, timestamp with time zone values are always normalized to UTC+0 regardless of the data source path. Here's why:
> 
> ### Streaming Path (pgoutput logical replication)
> When consuming WAL via the pgoutput plugin, the value goes through `AbstractColumnValue.asOffsetDateTimeAtUtc()` :
> 
> ```java
> return DateTimeFormat.get()
>     .timestampWithTimeZoneToOffsetDateTime(asString())   // Step 1: parse raw text with original offset
>     .withOffsetSameInstant(ZoneOffset.UTC);              // Step 2: normalize to UTC+0
> ```
> 
> The pgoutput protocol delivers column values as UTF-8 text strings (e.g., "2019-10-02T13:51:30.123456+02:00"). The parser first produces an OffsetDateTime with the original timezone offset from the WAL entry, then explicitly converts it to UTC via .withOffsetSameInstant(ZoneOffset.UTC). This guarantees the OffsetDateTime entering convertTimestampWithZone() is always at offset zero.
> 
> ### Snapshot Path (JDBC)
> During initial snapshots, the JDBC driver returns java.sql.Timestamp (a subclass of java.util.Date). In convertTimestampWithZone():
> 
> ```java
> if (data instanceof java.util.Date) {
>     data = OffsetDateTime.ofInstant(((Date) data).toInstant(), ZoneOffset.UTC);
> }
> ```
> 
> <img alt="image" width="796" height="538" src="https://private-user-images.githubusercontent.com/125648852/598008296-0af0542e-52d1-46cc-942d-3225c2906b64.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzk3Nzg3MDMsIm5iZiI6MTc3OTc3ODQwMywicGF0aCI6Ii8xMjU2NDg4NTIvNTk4MDA4Mjk2LTBhZjA1NDJlLTUyZDEtNDZjYy05NDJkLTMyMjVjMjkwNmI2NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNTI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDUyNlQwNjUzMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZTM4ZTVlZjY1ZDc5NDAwMGE5MjFmMGMxNWM2ZmY3MGNmN2FhMGYzZGVjZDUxY2MyYThlNDc1Mzk0YmFjYzRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.z2Gp0Bkuow4plre2ULSmrFTsIa7XeRxjqcjtgOzZ748">
> java.util.Date.toInstant() yields an absolute point in time (epoch-based), and OffsetDateTime.ofInstant(..., ZoneOffset.UTC) renders it explicitly at UTC+0.
> 
> Thus, using timestamp ltz is ok because time offset is normalized to UTC+0 by debezium. The only concern is compatibility, but it should be fine. Although the downstream jobs created previously used timestamp with time zone instead of timestamp with local time zone, no data has been written yet. Users will need to delete the downstream jobs and re-submit them. @leonardBang ,WDYT?

Thanks @loserwang1024 for the detailed analysis, I think there're no normal PG CDC job with timestamptz type in users' existing pipeline as the error will happen immediately once the DataChangEvent arrives. 

+1 to map PG’s timestamptz type to Flink CDC timestamp_ltz type.

loserwang1024 reviewed (approved):
LGTM!
